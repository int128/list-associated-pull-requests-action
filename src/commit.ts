type Commit = {
  commitId: string
  committedDate: Date
}

export const findEarliestCommit = (commits: Commit[]): Commit => {
  let earliestCommit = commits[0]
  for (const commit of commits) {
    if (commit.committedDate.getTime() < earliestCommit.committedDate.getTime()) {
      earliestCommit = commit
    }
  }
  return earliestCommit
}
