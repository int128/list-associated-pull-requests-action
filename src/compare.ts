import * as core from '@actions/core'
import { GitHub } from '@actions/github/lib/utils'

type Octokit = InstanceType<typeof GitHub>

type Inputs = {
  owner: string
  repo: string
  base: string
  head: string
}

type Outputs = {
  commitIds: Set<string>
  earliestCommitId: string
  earliestCommitDate: Date
}

export const compareCommits = async (octokit: Octokit, inputs: Inputs): Promise<Outputs> => {
  const compareIterator = octokit.paginate.iterator(octokit.rest.repos.compareCommitsWithBasehead, {
    owner: inputs.owner,
    repo: inputs.repo,
    basehead: `${inputs.base}...${inputs.head}`,
    per_page: 100,
  })

  const commitIds = new Set<string>()
  let earliestCommitDate = new Date()
  let earliestCommitId = ''
  for await (const compare of compareIterator) {
    core.info(`Received ${compare.data.commits.length} commits of total ${compare.data.total_commits}`)
    core.info(`Remaining rate limit: ${compare.headers['x-ratelimit-remaining'] ?? '?'}`)

    for (const commit of compare.data.commits) {
      commitIds.add(commit.sha)
      if (commit.commit.committer?.date) {
        const d = new Date(commit.commit.committer.date)
        if (d < earliestCommitDate) {
          earliestCommitDate = d
          earliestCommitId = commit.sha
        }
      }
    }
  }
  return { commitIds, earliestCommitId, earliestCommitDate }
}
