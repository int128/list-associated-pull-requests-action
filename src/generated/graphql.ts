import * as Types from './graphql-types';

export type AssociatedPullRequestsInCommitHistoryOfSubTreeQueryVariables = Types.Exact<{
  owner: Types.Scalars['String'];
  name: Types.Scalars['String'];
  expression: Types.Scalars['String'];
  path: Types.Scalars['String'];
  since: Types.Scalars['GitTimestamp'];
  historySize: Types.Scalars['Int'];
  historyAfter?: Types.InputMaybe<Types.Scalars['String']>;
}>;


export type AssociatedPullRequestsInCommitHistoryOfSubTreeQuery = { __typename?: 'Query', rateLimit?: { __typename?: 'RateLimit', cost: number } | null, repository?: { __typename?: 'Repository', object?: { __typename: 'Blob' } | { __typename: 'Commit', history: { __typename?: 'CommitHistoryConnection', totalCount: number, pageInfo: { __typename?: 'PageInfo', hasNextPage: boolean, endCursor?: string | null }, nodes?: Array<{ __typename?: 'Commit', oid: string, associatedPullRequests?: { __typename?: 'PullRequestConnection', nodes?: Array<{ __typename?: 'PullRequest', number: number, author?: { __typename?: 'Bot', login: string } | { __typename?: 'EnterpriseUserAccount', login: string } | { __typename?: 'Mannequin', login: string } | { __typename?: 'Organization', login: string } | { __typename?: 'User', login: string } | null } | null> | null } | null } | null> | null } } | { __typename: 'Tag' } | { __typename: 'Tree' } | null } | null };
