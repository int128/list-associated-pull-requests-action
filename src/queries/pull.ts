import { GitHub } from '@actions/github/lib/utils'
import { PullRequestCommitsQuery, PullRequestCommitsQueryVariables } from '../generated/graphql'

type Octokit = InstanceType<typeof GitHub>

const query = /* GraphQL */ `
  query pullRequestCommits($owner: String!, $name: String!, $number: Int!, $commitCursor: String) {
    rateLimit {
      cost
    }
    repository(owner: $owner, name: $name) {
      pullRequest(number: $number) {
        headRefOid
        commits(first: 100, after: $commitCursor) {
          totalCount
          pageInfo {
            hasNextPage
            endCursor
          }
          nodes {
            commit {
              oid
              committedDate
              associatedPullRequests(first: 1) {
                nodes {
                  number
                }
              }
            }
          }
        }
      }
    }
  }
`

export const getPullRequestCommits = async (
  o: Octokit,
  v: PullRequestCommitsQueryVariables,
  a: PullRequestCommitsQuery[] = []
): Promise<PullRequestCommitsQuery[]> => {
  const r = await o.graphql<PullRequestCommitsQuery>(query, v)
  const pageInfo = r.repository?.pullRequest?.commits.pageInfo
  if (pageInfo && pageInfo.hasNextPage && pageInfo.endCursor) {
    return await getPullRequestCommits(o, { ...v, commitCursor: pageInfo.endCursor }, [...a, r])
  }
  return [...a, r]
}
