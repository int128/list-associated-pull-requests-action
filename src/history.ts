import * as core from '@actions/core'
import { GitHub } from '@actions/github/lib/utils'
import { AssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './generated/graphql'
import { getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './queries/history'

type Octokit = InstanceType<typeof GitHub>

type Inputs = {
  owner: string
  repo: string
  ref: string
  path: string
  since: Date
  sinceCommitId: string
}

export type Commit = {
  commitId: string
  pull?: {
    number: number
    author: string
  }
}

export const getCommitHistory = async (octokit: Octokit, inputs: Inputs): Promise<Commit[]> => {
  const commits: Commit[] = []
  for (let afterCursor: string | undefined; ; ) {
    const page = await getCommitHistoryByCursor(octokit, inputs, afterCursor)
    commits.push(...page.commits)
    if (page.hasNextPage === false) {
      break
    }
    afterCursor = page.endCursor
  }
  return commits
}

type CommitHistoryPage = {
  commits: Commit[]
  hasNextPage: boolean
  endCursor?: string
}

const getCommitHistoryByCursor = async (
  octokit: Octokit,
  inputs: Inputs,
  afterCursor?: string
): Promise<CommitHistoryPage> => {
  core.startGroup(
    `AssociatedPullRequestsInCommitHistoryOfSubTreeQuery (${inputs.ref}, ${inputs.path}, ${inputs.since.toISOString()})`
  )
  const q = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
    owner: inputs.owner,
    name: inputs.repo,
    expression: inputs.ref,
    path: inputs.path,
    since: inputs.since,
    after: afterCursor,
  })
  core.info(JSON.stringify(q, undefined, 2))
  core.endGroup()

  return parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(q, inputs.sinceCommitId)
}

export const parseAssociatedPullRequestsInCommitHistoryOfSubTreeQuery = (
  q: AssociatedPullRequestsInCommitHistoryOfSubTreeQuery,
  sinceCommitId: string
): CommitHistoryPage => {
  if (q.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(q.repository?.object?.__typename)} !== Commit`)
  }

  const nodes = filterNodes(q, sinceCommitId)
  core.startGroup(`Found ${nodes.length} commit(s)`)
  const commits: Commit[] = []
  for (const node of nodes) {
    if (!node.associatedPullRequests?.nodes?.length) {
      core.info(`${node.oid} -> no pull request`)
      commits.push({ commitId: node.oid })
      continue
    }
    for (const pull of node.associatedPullRequests.nodes) {
      if (pull?.number === undefined) {
        continue
      }
      core.info(`${node.oid} -> #${pull.number}`)
      commits.push({
        commitId: node.oid,
        pull: {
          number: pull.number,
          author: pull.author?.login ?? '',
        },
      })
    }
  }
  core.endGroup()

  return {
    commits,
    hasNextPage: q.repository.object.history.pageInfo.hasNextPage,
    endCursor: q.repository.object.history.pageInfo.endCursor ?? undefined,
  }
}

const filterNodes = (q: AssociatedPullRequestsInCommitHistoryOfSubTreeQuery, sinceCommitId: string) => {
  if (q.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(q.repository?.object?.__typename)} !== Commit`)
  }
  const nodes = []
  for (const node of q.repository.object.history.nodes ?? []) {
    if (node == null) {
      continue
    }
    nodes.push(node)
    if (node.oid === sinceCommitId) {
      break
    }
  }
  return nodes
}
