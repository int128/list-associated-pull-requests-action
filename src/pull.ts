import * as core from '@actions/core'
import * as github from '@actions/github'
import { PullRequestCommitsQuery, CommitHistoryOfSubTreeQuery } from './generated/graphql'
import { getPullRequestCommits } from './queries/pull'
import { getCommitHistoryOfSubTree } from './queries/subtree'

type Inputs = {
  token: string
  pullRequest: number
  groupBySubPaths: string[]
}

type Outputs = {
  body: string
  associatedPullRequests: number[]
}

export const computeChangeSetOfPullRequest = async (inputs: Inputs): Promise<Outputs> => {
  const octokit = github.getOctokit(inputs.token)

  const commitsQueryList = await getPullRequestCommits(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    number: inputs.pullRequest,
  })
  core.startGroup(`Commit of pull request #${inputs.pullRequest}`)
  core.info(JSON.stringify(commitsQueryList, undefined, 2))
  core.endGroup()

  const earliestCommittedDate = findEarliestCommittedDate(commitsQueryList)
  if (earliestCommittedDate === undefined) {
    throw new Error(`internal error: earliestCommittedDate === undefined`)
  }
  core.info(`earliestCommittedDate = ${earliestCommittedDate.toISOString()}`)

  const headRefOid = findHeadRefOid(commitsQueryList)
  if (headRefOid === undefined) {
    throw new Error(`internal error: headRefOid === undefined`)
  }
  core.info(`headRefOid = ${headRefOid}`)

  const historyQueryList = []
  for (const path of inputs.groupBySubPaths) {
    const q = await getCommitHistoryOfSubTree(octokit, {
      owner: github.context.repo.owner,
      name: github.context.repo.repo,
      oid: headRefOid,
      path,
      since: earliestCommittedDate,
    })
    core.startGroup(`Commit history of sub tree: ${path}`)
    core.info(JSON.stringify(q, undefined, 2))
    core.endGroup()
    historyQueryList.push({ path, q })
  }

  const { commitsOfPathList, commitsOfOthers, associatedPullRequests } = calculate(commitsQueryList, historyQueryList)

  if (commitsOfPathList.length === 0) {
    return {
      body: formatCommits(commitsOfOthers).join('\n'),
      associatedPullRequests: [...associatedPullRequests],
    }
  }

  const body = []
  for (const { path, commits } of commitsOfPathList) {
    body.push(`### ${path}`)
    body.push(...formatCommits(commits))
  }
  if (commitsOfOthers.length > 0) {
    body.push(`### Others`)
    body.push(...formatCommits(commitsOfOthers))
  }

  return {
    body: body.join('\n'),
    associatedPullRequests: [...associatedPullRequests],
  }
}

const formatCommits = (commits: Commit[]): string[] => [
  ...new Set(commits.map((commit) => `- ${commit.pull ? `#${commit.pull}` : commit.commitId}`)),
]

export const calculate = (
  commitsQueryList: PullRequestCommitsQuery[],
  historyQueryList: { path: string; q: CommitHistoryOfSubTreeQuery }[]
) => {
  const commitsOfPull = parsePullRequestCommitsQueryList(commitsQueryList)

  const commitsOfPathList = []
  for (const { path, q } of historyQueryList) {
    const commitIds = parseCommitHistoryOfSubTreeQuery(q)

    // pick a commit which exists in the pull request
    const commits = []
    for (const commit of commitsOfPull) {
      if (commitIds.has(commit.commitId)) {
        commits.push(commit)
      }
    }
    if (commits.length > 0) {
      commitsOfPathList.push({ path, commits })
    }
  }

  const commitsOfOthersMap = new Map<CommitId, Commit>(commitsOfPull.map((commit) => [commit.commitId, commit]))
  for (const { commits } of commitsOfPathList) {
    for (const commit of commits) {
      commitsOfOthersMap.delete(commit.commitId)
    }
  }
  const commitsOfOthers = [...commitsOfOthersMap.values()]

  const associatedPullRequests = new Set<number>()
  for (const { pull } of commitsOfPull) {
    if (pull !== undefined) {
      associatedPullRequests.add(pull)
    }
  }

  return { commitsOfPathList, commitsOfOthers, associatedPullRequests }
}

const findEarliestCommittedDate = (queryList: PullRequestCommitsQuery[]) => {
  let earliestCommittedDate: Date | undefined
  for (const q of queryList) {
    for (const node of q.repository?.pullRequest?.commits?.nodes ?? []) {
      if (node?.commit === undefined) {
        continue
      }
      const committedDate = new Date(node.commit.committedDate)
      if (earliestCommittedDate === undefined || committedDate < earliestCommittedDate) {
        earliestCommittedDate = committedDate
      }
    }
  }
  return earliestCommittedDate
}

const findHeadRefOid = (queryList: PullRequestCommitsQuery[]) => queryList[0].repository?.pullRequest?.headRefOid

type CommitId = string
type Commit = {
  commitId: CommitId
  pull?: number
}

const parsePullRequestCommitsQueryList = (queryList: PullRequestCommitsQuery[]): Commit[] => {
  const commits: Commit[] = []
  for (const q of queryList) {
    for (const node of q.repository?.pullRequest?.commits?.nodes ?? []) {
      if (node?.commit.oid === undefined) {
        continue
      }
      if (!node.commit.associatedPullRequests?.nodes?.length) {
        core.info(`${node.commit.oid} -> none (${node.commit.committedDate})`)
        commits.push({ commitId: node.commit.oid })
        continue
      }
      for (const pull of node.commit.associatedPullRequests.nodes) {
        if (pull == null) {
          continue
        }
        core.info(`${node.commit.oid} -> #${pull.number} (${node.commit.committedDate})`)
        commits.push({ commitId: node.commit.oid, pull: pull.number })
      }
    }
  }
  return commits
}

const parseCommitHistoryOfSubTreeQuery = (q: CommitHistoryOfSubTreeQuery): Set<CommitId> => {
  if (q.repository?.object?.__typename !== 'Commit') {
    throw new Error(`internal error: __typename !== Commit`)
  }
  const commitIds = new Set<CommitId>()
  for (const node of q.repository.object.history.nodes ?? []) {
    if (node == null) {
      continue
    }
    commitIds.add(node.oid)
  }
  return commitIds
}
