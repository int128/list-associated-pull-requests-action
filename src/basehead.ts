import * as core from '@actions/core'
import * as github from '@actions/github'
import { ChangeSet, findChangeSet } from './history'
import { getCommit } from './queries/commit'
import { getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery } from './queries/history'

type Inputs = {
  token: string
  base: string
  head: string
  path: string
  groupBySubPaths: string[]
}

type Outputs = {
  body: string
  associatedPullRequests: number[]
}

export const computeChangeSetBetweenBaseHead = async (inputs: Inputs): Promise<Outputs> => {
  const octokit = github.getOctokit(inputs.token)

  const baseCommit = await getCommit(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    expression: inputs.base,
  })
  core.startGroup(`Base commit of ${inputs.base}`)
  core.info(JSON.stringify(baseCommit, undefined, 2))
  core.endGroup()
  if (baseCommit.repository?.object?.__typename !== 'Commit') {
    throw new Error(`unexpected typename ${String(baseCommit.repository?.object?.__typename)} !== Commit`)
  }

  const history = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
    owner: github.context.repo.owner,
    name: github.context.repo.repo,
    expression: inputs.head,
    path: inputs.path,
    since: baseCommit.repository.object.committedDate,
  })
  core.startGroup(`Commit history on ${inputs.head} since ${baseCommit.repository.object.committedDate}`)
  core.info(JSON.stringify(history, undefined, 2))
  core.endGroup()
  const changeSet = findChangeSet(history, baseCommit.repository.object.oid)

  if (inputs.groupBySubPaths.length === 0) {
    return {
      body: [...changeSet.pullOrCommits].map((s) => `- ${s}`).join('\n'),
      associatedPullRequests: [...changeSet.pulls],
    }
  }

  const subPathChangeSets = new Map<string, ChangeSet>()
  for (const subPath of inputs.groupBySubPaths) {
    const history = await getAssociatedPullRequestsInCommitHistoryOfSubTreeQuery(octokit, {
      owner: github.context.repo.owner,
      name: github.context.repo.repo,
      expression: inputs.head,
      path: subPath,
      since: baseCommit.repository.object.committedDate,
    })
    core.startGroup(`Commit history of subpath ${subPath}`)
    core.info(JSON.stringify(history, undefined, 2))
    core.endGroup()
    const changeSet = findChangeSet(history, baseCommit.repository.object.oid)
    subPathChangeSets.set(subPath, changeSet)
  }

  const otherPullOrCommits = new Set(changeSet.pullOrCommits)
  const body = []
  for (const [subPath, { pullOrCommits }] of subPathChangeSets) {
    if (pullOrCommits.size > 0) {
      body.push(`### ${subPath}`)
      for (const pullOrCommit of pullOrCommits) {
        body.push(`- ${pullOrCommit}`)
        otherPullOrCommits.delete(pullOrCommit)
      }
    }
  }
  if (otherPullOrCommits.size > 0) {
    body.push(`### Others`)
    for (const pullOrCommit of otherPullOrCommits) {
      body.push(`- ${pullOrCommit}`)
    }
  }

  return {
    body: body.join('\n'),
    associatedPullRequests: [...changeSet.pulls],
  }
}
