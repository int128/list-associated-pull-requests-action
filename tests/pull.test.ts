import { CommitHistoryOfSubTreeQuery, PullRequestCommitsQuery } from '../src/generated/graphql'
import { calculate } from '../src/pull'

test('calculate', () => {
  // commit of pull request #98
  const commitsQueryList: PullRequestCommitsQuery[] = [
    {
      rateLimit: {
        cost: 1,
      },
      repository: {
        pullRequest: {
          headRefOid: '0f9bbfade221f8b590163c664e27c3839fffd3c8',
          commits: {
            totalCount: 26,
            pageInfo: {
              hasNextPage: false,
              endCursor: 'MjY',
            },
            nodes: [
              {
                commit: {
                  oid: '3d9cb2f7a130e4427676df5f6da9c06cd228d7d7',
                  committedDate: '2022-05-14T01:46:46Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 67,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '0ad635a9b2be184a6a02253bcdb090dab6e195c9',
                  committedDate: '2022-05-14T05:31:47Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 68,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'f656f6812207d5301492cab12f399463e35fbdbb',
                  committedDate: '2022-05-14T07:41:25Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 69,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '4449d60e7f1ce5fd3d26750c66b5d708cbac1d20',
                  committedDate: '2022-05-14T12:42:37Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 71,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'ead2316173fbd364d6d241b8c05df4a349c7821f',
                  committedDate: '2022-05-14T12:45:14Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 70,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '6c696b80a95830b40ae8ed882bbb9bbe1972b24c',
                  committedDate: '2022-05-14T12:52:59Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 73,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '1a74c41db6587e71a6c97a2a64741d6152a88638',
                  committedDate: '2022-05-15T01:04:29Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 74,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'b256f3b36b635040e4ce150eec4c4d2fdab3495c',
                  committedDate: '2022-05-15T07:01:51Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 75,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '797d211a801b403772b703ed29e60dc148d708b8',
                  committedDate: '2022-05-15T07:36:12Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 76,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '15a4eb0b903195d92d089454ab9ffa1c7598d3ef',
                  committedDate: '2022-05-15T08:30:44Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 77,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'e1d47057c2a2eafd8ebb570c8f7a4a84dffb234e',
                  committedDate: '2022-05-15T09:27:30Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 78,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'd5414767c25010a8d42f44121dba0ee870c1da53',
                  committedDate: '2022-05-16T20:30:24Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 79,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '49a339c426a28b6e5abd3203fadcc749560b2362',
                  committedDate: '2022-05-17T03:39:16Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 80,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '6df54289158cbd5a61c8604cf97b8a0f7d1847f7',
                  committedDate: '2022-05-17T18:43:59Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 81,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'd486e046fca314b3a42693a5ca878e47c69a89ed',
                  committedDate: '2022-05-21T03:07:41Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 82,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '13868898d92652c27d889259bd11215f7a4551c1',
                  committedDate: '2022-05-23T21:43:39Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 83,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'aed841c2c1550f2e2bac160837085856b6cc5120',
                  committedDate: '2022-05-24T11:59:55Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 84,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '8e07a09848e1158d3326ec2ab9006077c087bd33',
                  committedDate: '2022-05-24T22:39:38Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 85,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'fc7d55b8f3e86b9f2b71c30d9e8578dffc40c807',
                  committedDate: '2022-05-26T02:26:42Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 86,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'ebb87b274ad05f7e5d9a38c677f2e399041dae01',
                  committedDate: '2022-05-26T09:52:50Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 87,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'ca85088adb0f411ef934833e2b4c47941794ac58',
                  committedDate: '2022-05-27T00:20:11Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 88,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'aba1fb9e861f34176727a527a6e5af8ba74f628c',
                  committedDate: '2022-05-27T00:29:05Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 89,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '4e216f67567bdbb723ad28b1d946c63b6cb09a8d',
                  committedDate: '2022-05-28T05:09:40Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 91,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'dcff886579270638a2146cf8ba5380f2240774c7',
                  committedDate: '2022-05-28T12:28:25Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 96,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: 'f2ac78bee33a1a55186732a1f2fb6b7d68008a53',
                  committedDate: '2022-05-28T14:34:18Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 90,
                      },
                    ],
                  },
                },
              },
              {
                commit: {
                  oid: '0f9bbfade221f8b590163c664e27c3839fffd3c8',
                  committedDate: '2022-05-28T23:41:47Z',
                  associatedPullRequests: {
                    nodes: [
                      {
                        number: 97,
                      },
                    ],
                  },
                },
              },
            ],
          },
        },
      },
    },
  ]
  const historyQueryList: { path: string; q: CommitHistoryOfSubTreeQuery }[] = [
    {
      path: 'src',
      q: {
        rateLimit: {
          cost: 1,
        },
        repository: {
          object: {
            __typename: 'Commit',
            history: {
              nodes: [
                {
                  oid: 'aba1fb9e861f34176727a527a6e5af8ba74f628c',
                },
                {
                  oid: 'ca85088adb0f411ef934833e2b4c47941794ac58',
                },
                {
                  oid: 'ebb87b274ad05f7e5d9a38c677f2e399041dae01',
                },
                {
                  oid: 'e1d47057c2a2eafd8ebb570c8f7a4a84dffb234e',
                },
                {
                  oid: '797d211a801b403772b703ed29e60dc148d708b8',
                },
                {
                  oid: 'b256f3b36b635040e4ce150eec4c4d2fdab3495c',
                },
                {
                  oid: 'ead2316173fbd364d6d241b8c05df4a349c7821f',
                },
                {
                  oid: 'f656f6812207d5301492cab12f399463e35fbdbb',
                },
                {
                  oid: '0ad635a9b2be184a6a02253bcdb090dab6e195c9',
                },
              ],
            },
          },
        },
      },
    },
    {
      path: '.github',
      q: {
        rateLimit: {
          cost: 1,
        },
        repository: {
          object: {
            __typename: 'Commit',
            history: {
              nodes: [
                {
                  oid: 'dcff886579270638a2146cf8ba5380f2240774c7',
                },
                {
                  oid: '4e216f67567bdbb723ad28b1d946c63b6cb09a8d',
                },
                {
                  oid: 'ebb87b274ad05f7e5d9a38c677f2e399041dae01',
                },
                {
                  oid: 'e1d47057c2a2eafd8ebb570c8f7a4a84dffb234e',
                },
                {
                  oid: 'ead2316173fbd364d6d241b8c05df4a349c7821f',
                },
              ],
            },
          },
        },
      },
    },
  ]
  const { commitsOfPathList, commitsOfOthers, associatedPullRequests } = calculate(commitsQueryList, historyQueryList)
  expect(associatedPullRequests).toMatchSnapshot()
  expect(commitsOfPathList).toMatchSnapshot()
  expect(commitsOfOthers).toMatchSnapshot()
})
