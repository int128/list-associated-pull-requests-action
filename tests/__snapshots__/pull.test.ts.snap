// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate 1`] = `
Set {
  67,
  68,
  69,
  71,
  70,
  73,
  74,
  75,
  76,
  77,
  78,
  79,
  80,
  81,
  82,
  83,
  84,
  85,
  86,
  87,
  88,
  89,
  91,
  96,
  90,
  97,
}
`;

exports[`calculate 2`] = `
Array [
  Object {
    "commits": Array [
      Object {
        "commitId": "0ad635a9b2be184a6a02253bcdb090dab6e195c9",
        "pull": 68,
      },
      Object {
        "commitId": "f656f6812207d5301492cab12f399463e35fbdbb",
        "pull": 69,
      },
      Object {
        "commitId": "ead2316173fbd364d6d241b8c05df4a349c7821f",
        "pull": 70,
      },
      Object {
        "commitId": "b256f3b36b635040e4ce150eec4c4d2fdab3495c",
        "pull": 75,
      },
      Object {
        "commitId": "797d211a801b403772b703ed29e60dc148d708b8",
        "pull": 76,
      },
      Object {
        "commitId": "e1d47057c2a2eafd8ebb570c8f7a4a84dffb234e",
        "pull": 78,
      },
      Object {
        "commitId": "ebb87b274ad05f7e5d9a38c677f2e399041dae01",
        "pull": 87,
      },
      Object {
        "commitId": "ca85088adb0f411ef934833e2b4c47941794ac58",
        "pull": 88,
      },
      Object {
        "commitId": "aba1fb9e861f34176727a527a6e5af8ba74f628c",
        "pull": 89,
      },
    ],
    "path": "src",
  },
  Object {
    "commits": Array [
      Object {
        "commitId": "ead2316173fbd364d6d241b8c05df4a349c7821f",
        "pull": 70,
      },
      Object {
        "commitId": "e1d47057c2a2eafd8ebb570c8f7a4a84dffb234e",
        "pull": 78,
      },
      Object {
        "commitId": "ebb87b274ad05f7e5d9a38c677f2e399041dae01",
        "pull": 87,
      },
      Object {
        "commitId": "4e216f67567bdbb723ad28b1d946c63b6cb09a8d",
        "pull": 91,
      },
      Object {
        "commitId": "dcff886579270638a2146cf8ba5380f2240774c7",
        "pull": 96,
      },
    ],
    "path": ".github",
  },
]
`;

exports[`calculate 3`] = `
Array [
  Object {
    "commitId": "3d9cb2f7a130e4427676df5f6da9c06cd228d7d7",
    "pull": 67,
  },
  Object {
    "commitId": "4449d60e7f1ce5fd3d26750c66b5d708cbac1d20",
    "pull": 71,
  },
  Object {
    "commitId": "6c696b80a95830b40ae8ed882bbb9bbe1972b24c",
    "pull": 73,
  },
  Object {
    "commitId": "1a74c41db6587e71a6c97a2a64741d6152a88638",
    "pull": 74,
  },
  Object {
    "commitId": "15a4eb0b903195d92d089454ab9ffa1c7598d3ef",
    "pull": 77,
  },
  Object {
    "commitId": "d5414767c25010a8d42f44121dba0ee870c1da53",
    "pull": 79,
  },
  Object {
    "commitId": "49a339c426a28b6e5abd3203fadcc749560b2362",
    "pull": 80,
  },
  Object {
    "commitId": "6df54289158cbd5a61c8604cf97b8a0f7d1847f7",
    "pull": 81,
  },
  Object {
    "commitId": "d486e046fca314b3a42693a5ca878e47c69a89ed",
    "pull": 82,
  },
  Object {
    "commitId": "13868898d92652c27d889259bd11215f7a4551c1",
    "pull": 83,
  },
  Object {
    "commitId": "aed841c2c1550f2e2bac160837085856b6cc5120",
    "pull": 84,
  },
  Object {
    "commitId": "8e07a09848e1158d3326ec2ab9006077c087bd33",
    "pull": 85,
  },
  Object {
    "commitId": "fc7d55b8f3e86b9f2b71c30d9e8578dffc40c807",
    "pull": 86,
  },
  Object {
    "commitId": "f2ac78bee33a1a55186732a1f2fb6b7d68008a53",
    "pull": 90,
  },
  Object {
    "commitId": "0f9bbfade221f8b590163c664e27c3839fffd3c8",
    "pull": 97,
  },
]
`;
